File: 105-Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(preorder.empty() || inorder.empty()) return NULL;
        return buildSubTree(preorder, inorder, preorder.size(), 0, 0);
    }
    
    TreeNode* buildSubTree(vector<int>& preorder, vector<int>& inorder, int length, int preStart, int inStart) {
        if(length==0) return NULL;
        if(length==1) return new TreeNode(preorder[preStart]);
        TreeNode *node = new TreeNode(preorder[preStart]);
        int i = inStart;
        while(inorder.at(i)!=preorder[preStart]){
            i++;
        }
        node->left = buildSubTree(preorder, inorder, i-inStart, preStart+1, inStart);
        node->right = buildSubTree(preorder, inorder, length+inStart-i-1, preStart-inStart+i+1, i+1);
        return node;
    }
};

File: 105-Construct-Binary-Tree-from-Preorder-and-Inorder-Traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder == null || inorder == null || preorder.length != inorder.length || preorder.length == 0) {
            return null;
        }
        return buildTree(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }

    private TreeNode buildTree(int[] preorder, int pre_start, int pre_end,
                               int[] inorder, int in_start, int in_end) {
        if (pre_start > pre_end || in_start > in_end) {
            return null;
        }
        TreeNode node = new TreeNode(preorder[pre_start]);
        if (pre_start < pre_end) {
            int size = 0;
            while (in_start + size < in_end && preorder[pre_start] != inorder[in_start + size]) {
                size++;
            }
            node.left = buildTree(preorder, pre_start + 1, pre_start + size, inorder, in_start, in_start + size - 1);
            node.right = buildTree(preorder, pre_start + 1 + size, pre_end, inorder, in_start + size + 1, in_end);
        }
        return node;
    }
}